Add mstest runner docs (#38459)
dotnet/docs: docs/core/testing/unit-testing-mstest-runner-intro.md

---
title: MSTest Runner
description: Learn about MSTest Runner, a lightweight way to run tests without depending on dotnet SDK.
author: nohwnd
ms.author: jajares
ms.date: 11/28/2023
ms.custom: 
---

# Overview

> MSTest Runner is a light-weight and portable alternative to [microsoft/vstest](https://github.com/microsoft/vstest) for running tests in continuous integration pipelines, and in VisualStudio TestExplorer (Coming soon!).

Author: ```suggestion
MSTest Runner is a light-weight and portable alternative to [microsoft/vstest](https://github.com/microsoft/vstest) for running tests in continuous integration pipelines, and in VisualStudio TestExplorer (as a Preview Feature).
```

---
title: MSTest runner overview
description: Learn about the MSTest runner, a lightweight way to run tests without depending on the .NET SDK.
author: nohwnd
ms.author: jajares
ms.date: 11/28/2023
---

# MSTest runner overview

The MSTest runner is a lightweight and portable alternative to [VSTest](https://github.com/microsoft/vstest) for running tests in continuous integration (CI) pipelines, and in Visual Studio Test Explorer.

The MSTest runner is embedded directly in your MSTest test projects, and there's no additional app dependencies, such as `vstest.console` or `dotnet test` needed to run your tests.

> The MSTest runner open source and you can view its code in the [microsoft/testfx](https://github.com/microsoft/testfx/tree/main/src/Platform/Microsoft.Testing.Platform) GitHub repository. The MSTest runner comes bundled with MSTest in 3.2.0-preview. This preview is available in the [test-tools NuGet feed](https://pkgs.dev.azure.com/dnceng/public/_packaging/test-tools/nuget/v3/index.json).

Author: ```suggestion
The MSTest runner is open source, and builds on a Microsoft.Testing.Platform library. You can find Microsoft.Testing.Platform code in [microsoft/testfx](https://github.com/microsoft/testfx/tree/main/src/Platform/Microsoft.Testing.Platform) GitHub repository. The MSTest runner comes bundled with MSTest in 3.2.0-preview. This preview is available in the [test-tools NuGet feed](https://pkgs.dev.azure.com/dnceng/public/_packaging/test-tools/nuget/v3/index.json).
```

### Use `dotnet run`

The `dotnet run` command can be used to build and run your test project. This is the easiest, although sometimes slowest, way to run your tests. Using `dotnet run` is practical when you are editing and running tests locally, because it ensures that the test project is re-built when needed. `dotnet run` will also automatically find the project in the current folder.

### Use `dotnet exec`

`dotnet exec` can be used to run an already built test project, this is an alternative to running the application directly. `dotnet exec` requires path to the built test project dll.

> [!IMPORTANT]
> Providing the path to the test project executable (_*.exe_) results in an error:
>
> ```Output
> Error:
>   An assembly specified in the application dependencies manifest
>   (Contoso.MyTests.deps.json) has already been found but with a different
>   file extension:
>     package: 'Contoso.MyTests', version: '1.0.0'
>     path: 'Contoso.MyTests.dll'
>     previously found assembly: 'S:\t\Contoso.MyTests\bin\Debug\net8.0\Contoso.MyTests.exe'
> ```

#### Use `dotnet test`

> > [!IMPORTANT]
> The `dotnet test` command only works with MSTest, NUnit and xUnit tests.

> For tests authored with MSTest, NUnit or xUnit test framework, it's possible to run tests with `dotnet test`.

Author: ```suggestion
For tests authored with MSTest, NUnit or xUnit test framework, it's possible to run tests with [dotnet test](https://learn.microsoft.com/dotnet/core/tools/dotnet-test).
```

---
title: MSTest Runner
description: Learn about MSTest Runner, a lightweight way to run tests without depending on dotnet SDK.
author: nohwnd
ms.author: jajares
ms.date: 11/28/2023
ms.custom: 
---

# Overview

MSTest Runner is a light-weight and portable alternative to [microsoft/vstest](https://github.com/microsoft/vstest) for running tests in continuous integration pipelines, and in VisualStudio TestExplorer (as a Preview Feature).

> This runner is embedded directly in your MSTest test project, and there is no additional application (e.g. vstest.console / dotnet test) or any additional infrastructure (e.g. dotnet SDK) needed to run your tests.

Reviewer: @IEvangelist With `dotnet test` you do need the SDK to be installed. With `vstest.console.exe` (wrapped by `dotnet test`) you need Visual Studio, Visual Studio Build Tools or the Test Platform tool package installed. Whereas with this solution you don't need anything extra.

Also, it's important to mention that on the contrary of VSTest, with this runner, you can now generate dotnet self-contained app that also don't need .NET runtime to be installed. You can simply copy the output folder and move it to a bare machine to be able to run your tests.

> This runner is embedded directly in your MSTest test project, and there is no additional application (e.g. vstest.console / dotnet test) or any additional infrastructure (e.g. dotnet SDK) needed to run your tests.

Reviewer 2: Hi @Evangelink, thanks for the reply here. I still have questions... 🤓

But wouldn't you still need the .NET SDK? That's my concern, if you're authoring unit tests and trying to run them, you're going to need the .NET SDK to do so. You can't build the exe without the .NET SDK. The same is true for when you're running as a CI/CD pipeline, like GitHub Actions for example, to run the tests, you'll presumably need the .NET SDK to build the exe and then run it.

> This runner is embedded directly in your MSTest test project, and there is no additional application (e.g. vstest.console / dotnet test) or any additional infrastructure (e.g. dotnet SDK) needed to run your tests.

Reviewer: You need SDK to build the test executable but you don't need it to run the tests. With current solution (VSTest) even if you have produced your dll, you DO need the SDK to run `dotnet test Contoso.dll`. With this solution, you can simply call `Contoso.exe`

```csproj
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <!-- The secret sauce to enabling MSTestRunner. -->
    <UseMSTestRunner>true</UseMSTestRunner>

    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>

    <IsPackable>false</IsPackable>
    <IsTestProject>true</IsTestProject>
  </PropertyGroup>

  <ItemGroup>
    <!-- Replace these 3 packages by reference to just MSTest 
      <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.7.2" />
      <PackageReference Include="MSTest.TestAdapter" Version="3.2.0-preview.23570.1" />
      <PackageReference Include="MSTest.TestFramework" Version="3.2.0-preview.23570.1" />
    -->

    <PackageReference Include="MSTest" Version="3.2.0-preview.23570.1" />

    <!-- Replace coverlet collector 
      <PackageReference Include="coverlet.collector" Version="6.0.0" />
    --> 
    <PackageReference Include="Microsoft.Testing.Platform.Extensions.CodeCoverage" Version="17.9.4-beta.23563.1" />
  </ItemGroup>

</Project>
```

> Building the application will generate an executable application for your test project. For example `Contoso.MyTests.exe` on Windows.

Reviewer: We want to make it explicit that compared to current testing solution that produces dll that needs something to run them, with the runner you are directly producing a test application as a regular executable.

#### Using `dotnet exec`

`dotnet exec` can be used to run an already built test project, this is an alternative to running the application directly. `dotnet exec` requires path to the built test project dll.

> [!IMPORTANT]
> Providing path to the test project executable (.exe) will result in an error:

> ```plaintext
> Error:
>   An assembly specified in the application dependencies manifest
>   (Contoso.MyTests.deps.json) has already been found but with a different
>   file extension:
>     package: 'Contoso.MyTests', version: '1.0.0'
>     path: 'Contoso.MyTests.dll'
>     previously found assembly: 'S:\t\Contoso.MyTests\bin\Debug\net8.0\Contoso.MyTests.exe'
> ```

#### Using `dotnet test`

> > [!IMPORTANT]
> `dotnet test` is currently working only with MSTest, NUnit and xUnit tests.

For tests authored with MSTest, NUnit or xUnit test framework, it is possible to run tests with `dotnet test`.

Provide a path to the tested dll, or to the project and your tests will run.

### In Visual Studio

> Testing.Platform tests can be run (and debugged) in VisualStudio, they integrate with Test Explorer, and can also be run directly as Startup Project.

Reviewer: I am not sure how to best word that. The idea is to differentiate MSTest tests produced the old way to test produced as self-contained test app.

Besides, MSTest runner is based on Microsoft.Testing.Platform solution that we hope to offer as alternative to VSTest also for xUnit and NUnit but we are waiting to get enough feedback on MSTest side before opening it further.

> Testing.Platform tests can be run (and debugged) in VisualStudio, they integrate with Test Explorer, and can also be run directly as Startup Project.

Author: I've added a suggestion of a quick note about Microsoft.Testing.Platform to the first paragraph. But it feels like there should be more explanations about what it is. 

I am also not sure how to properly let the user know that this is not an article about vstest.

> Testing.Platform tests can be run (and debugged) in VisualStudio, they integrate with Test Explorer, and can also be run directly as Startup Project.

Author: Maybe adding another article that describes vstest vs testing.platform, and call out the namespaces used, and fundamental differences?

> Testing.Platform tests can be run (and debugged) in VisualStudio, they integrate with Test Explorer, and can also be run directly as Startup Project.

Author: Pushed the basics of what I think should be covered in a new file in this PR. But it is hard for me to judge what to put there , because I am so close to the project.

#### Running the application directly in VS

Testing.Platform test project are built as executables, and can be run directly. This will run all the tests in the given executable, unless a filter is provided.

1. Navigate the test project you want to run in Solution Explorer, right click it and select `Set as Startup Project`.

> 1. Go to `Debug > Start without Debugging`, (or press `Ctrl + F5`) to run the selected test project.

Reviewer 2: ```suggestion
1. Select **Debug** > **Start without Debugging** (or use <kbd>Ctrl</kbd>+<kbd>F5</kbd>) to run the selected test project.
```

---
title: MSTest Test Runner
description: Learn about MSTest Test Runner, a lightweight way to run tests without depending on dotnet SDK.
author: nohwnd
ms.author: jajares
ms.date: 11/28/2023
ms.custom: 
---

# Overview

> MSTestRunner is a light-weight and portable alternative to [microsoft/vstest](https://github.com/microsoft/vstest) for running tests in continuous integration pipelines, and in VisualStudio TestExplorer (Coming soon!).

Author: ```suggestion
MSTest Runner is a light-weight and portable alternative to [microsoft/vstest](https://github.com/microsoft/vstest) for running tests in continuous integration pipelines, and in VisualStudio TestExplorer (Coming soon!).
```

> MSTestRunner is a light-weight and portable alternative to [microsoft/vstest](https://github.com/microsoft/vstest) for running tests in continuous integration pipelines, and in VisualStudio TestExplorer (Coming soon!).

Author: changed wording.

This runner is embedded directly in your MSTest test project, and there is no additional application (e.g. vstest.console / dotnet test) or any additional infrastructure (e.g. dotnet SDK) needed to run your tests.

MSTestRunner lives in [microsoft/testfx](https://github.com/microsoft/testfx/tree/main/src/Platform/Microsoft.Testing.Platform) repository, and comes bundled with MSTest in 3.2.0-preview. This preview is available on [test-tools NuGet feed](https://pkgs.dev.azure.com/dnceng/public/_packaging/test-tools/nuget/v3/index.json).

> ## Enabling MSTest Runner in an MSTest project

Reviewer: Shall it be `a MSTest` instead of `an MSTest`?

Author: ```suggestion
## Enabling MSTest Runner in a MSTest project
```

To enable MSTest Runnerin an MSTest project, you need to add `<UseMSTestRunner>true</UseMSTestRunner>` into your project file, and make sure you are using MSTest 3.2.0-preview or newer:

```csproj
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <!-- The secret sauce to enabling MSTestRunner. -->
    <UseMSTestRunner>true</UseMSTestRunner>

    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>

    <IsPackable>false</IsPackable>
    <IsTestProject>true</IsTestProject>
  </PropertyGroup>

  <ItemGroup>
    <!-- All these 3 packages can be replaced by reference to:

>         <PackageReference Include="MSTest" Version="3.2.0-preview.23570.1" />

Author: done.

    -->
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.7.2" />
    <PackageReference Include="MSTest.TestAdapter" Version="3.2.0-preview.23570.1" />
    <PackageReference Include="MSTest.TestFramework" Version="3.2.0-preview.23570.1" />

>     <PackageReference Include="coverlet.collector" Version="6.0.0" />

Author: done. And replaced by our CC.

  </ItemGroup>

</Project>
```

> Building the application will generate an executable application for your test project. For example `Contoso.MyTests.exe`.

Author: ```suggestion
Building the application will generate an executable application for your test project. For example `Contoso.MyTests.exe` on Windows.
```

#### Using `dotnet run`

`dotnet run` can be used to build and run your test project. This is the easiest, although sometimes slow, way to run your tests. Using `dotnet run` is practical when you are editing and running tests locally because it ensures that the test project is re-built when needed. `dotnet run` will also automatically find the project in the current folder.

#### Using `dotnet exec`

`dotnet exec` can be used to run an already built test project, this is an alternative to running the application directly. `dotnet exec` requires path to the built test project dll.

> [!IMPORTANT]
> Providing path to the test project executable (.exe) will result in an error:

> ```plaintext
> Error:
>   An assembly specified in the application dependencies manifest
>   (Contoso.MyTests.deps.json) has already been found but with a different
>   file extension:
>     package: 'Contoso.MyTests', version: '1.0.0'
>     path: 'Contoso.MyTests.dll'
>     previously found assembly: 'S:\t\Contoso.MyTests\bin\Debug\net8.0\Contoso.MyTests.exe'
> ```

#### Using `dotnet test`

> > [!IMPORTANT]

> > `dotnet test` is currently working only with MSTest, NUnit and xUnit tests. TPTF tests will not run using `dotnet test` and no error will be reported.

Reviewer: I don't think we should mention TPTF as it's only developed to test the platform but won't be delivered on its own.

Author: ```suggestion
> `dotnet test` is currently working only with MSTest, NUnit and xUnit tests.
```

For tests authored with MSTest, NUnit or xUnit test framework, it is possible to run tests with `dotnet test`.

> Provide a path to the tested dll, or to the project and your tests will run.

Reviewer: Might be worth to add a link to the `dotnet test` docs.

> ### In Visual Studio

Reviewer: We should have a specific section about VS Code

Author: Yes. I am not able to write it right now.

Testing.Platform tests can be run (and debugged) in VisualStudio, they integrate with Test Explorer, and can also be run directly as Startup Project.

#### Running the application directly in VS

Testing.Platform test project are built as executables, and can be run directly. This will run all the tests in the given executable, unless a filter is provided.

1. Navigate the test project you want to run in Solution Explorer, right click it and select `Set as Startup Project`.

> 1. Go to Debug > Start without Debugging, (or press `Ctrl + F5`) to run the selected test project.

Reviewer: ```suggestion
1. Go to `Debug > Start without Debugging`, (or press `Ctrl + F5`) to run the selected test project.
```

Console window will pop up with the execution and summary of your test run.

#### Debugging the application directly in VS

Testing.Platform test project are built as executables, and can be debugged directly. This will debug all the tests in the given executable, unless a filter is provided.

1. Navigate the test project you want to run in Solution Explorer, right click it and select `Set as Startup Project`.

1. Set breakpoint into the test that you'd like to debug.

> 1. Go to Debug > Start Debugging, (or press `F5`) to debug the selected test project.

Reviewer: ```suggestion
1. Go to `Debug > Start Debugging`, (or press `F5`) to debug the selected test project.
```

All tests will be executed until your test with a breakpoint is reached. Step through your test to debug it. Once you are done debugging the application will resume running all remaining tests, unless you stop it.

### Using Test Explorer

Testing.Platform tests partially integrate with Test Explorer. Tests can be Run and Debugged from Test Explorer.

> [!IMPORTANT]
> Automatic update of tests without building the project is not yet available.

To run a test, navigate to Test Explorer, select the test (or tests) to run. Right click it, and choose `Run`.

Similarly to debug a test, select the test (or tests) in Test Explorer, right click and choose `Debug`.

## In Continuous Integration (CI)

There is no special pipeline task, or any additional tooling to run Testing.Platform tests. There is also no additional tooling to run multiple tests projects through a single command.

To run a test project in CI add one step for each test executable that you wish to run, such as this on AzureDevOps:

```yml
- task: CmdLine@2
  displayName: "Run Contonso.MyTests.exe"
  inputs:
    script: '.\Contoso.MyTests\bin\Debug\net8.0\Contoso.MyTests.exe'
```

> ## Test Reports

Reviewer: I would group this and the following entries under a `## Extensions` section

Author: You probably meant also the Troubleshooting part, but I like that it is separate, but see for yourself and do as you please :)

**Environment variables take precedence on the command line arguments.**

### Hang Dump

This extension allows to create a dump file after a given timeout.

It can be configured using the following options:

| Option | Description |
| ------ | ----------- |
| `--hangdump` | Generated a dump file in case of test host process hang. |
| `-⁠-⁠hangdump-⁠filename` | Specify the file name of the dump. |
| `--hangdump-timeout` | Specify the timeout after which the dump will be generated. The timeout value is specified in one of the following formats:<br/>1.5h, 1.5hour, 1.5hours<br/>90m, 90min, 90minute, 90minutes<br/>5400s, 5400sec, 5400second, 5400seconds. Default is 30m. |
| `--hangdump-type` | Specify the type of the dump. Valid values are `Mini`, `Heap`, `Triage` (only .NET 6+), `Full`. Default type is `Full`. For more information visit [MS Learn Crash Dump Types](https://learn.microsoft.com/dotnet/core/diagnostics/collect-dumps-crash#types-of-mini-dumps) |

> To be able to use this feature you will need to
> 
> 1. Add a dependency to `Microsoft.Testing.Platform.Extensions.HangDump`
> 
> 2. Update your `Program.cs` to add
> 
>     ```csharp
>     using Microsoft.Testing.Platform.Extensions;
>     builder.AddHangDumpGenerator();
>     ```

Reviewer: No longer relevant, simply adding the package does everything automatically.

```suggestion
```

### Crash Dump

This extension allows to create a crash dump file in case of process crash.

It can be configured using the following options:

| Option | Description |
| ------ | ----------- |
| `--crashdump` | Generated a dump file in case of test host process crash. Supported by .NET 6.0+ |
| `⁠-⁠-⁠crashdump-⁠filename` | Specify the file name of the dump. |
| `--crashdump-type` | Specify the type of the dump. Valid values are `Mini`, `Heap`, `Triage`, `Full`. Default type is `Full`. For more information visit [MS Learn Crash Dump Types](https://learn.microsoft.com/dotnet/core/diagnostics/collect-dumps-crash#types-of-mini-dumps) |

> To be able to use this feature you will need to
> 
> 1. Add a dependency to `Microsoft.Testing.Platform.Extensions`
> 
> 1. Update your `Program.cs` to add
> 
>     ```csharp
>     using Microsoft.Testing.Platform.Extensions;
>     builder.AddCrashDumpGenerator();
>     ```

Reviewer: No longer relevant

```suggestion
```
